program-test/tests/sysvar_last_restart_slot.rs

```diff
@@ -0,0 +1,104 @@
+use {
+    solana_program_test::{processor, ProgramTest, ProgramTestContext},
+    solana_sdk::{
+        account_info::AccountInfo,
+        clock::Slot,
+        entrypoint::ProgramResult,
+        instruction::{AccountMeta, Instruction},
+        msg,
+        pubkey::Pubkey,
+        signature::Signer,
+        sysvar::{last_restart_slot, last_restart_slot::LastRestartSlot, Sysvar},
+        transaction::Transaction,
+    },
+};
+
+// program to check both syscall and sysvar
+fn sysvar_last_restart_slot_process_instruction(
+    _program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    input: &[u8],
+) -> ProgramResult {
+    msg!("sysvar_last_restart_slot");
+    assert_eq!(input.len(), 8);
+    let expected_last_hardfork_slot = u64::from_le_bytes(input[0..8].try_into().unwrap());
+
+    let last_restart_slot = LastRestartSlot::get();
+    msg!("last restart slot: {:?}", last_restart_slot);
+    assert_eq!(
+        last_restart_slot,
+        Ok(LastRestartSlot {
+            last_restart_slot: expected_last_hardfork_slot
+        })
+    );
+
+    let last_restart_slot_account = &accounts[0];
+    let slot_via_account = LastRestartSlot::from_account_info(last_restart_slot_account)?;
+    msg!("slot via account: {:?}", slot_via_account);
+
+    assert_eq!(
+        slot_via_account,
+        LastRestartSlot {
+            last_restart_slot: expected_last_hardfork_slot
+        }
+    );
+
+    Ok(())
+}
+
+async fn check_with_program(
+    context: &mut ProgramTestContext,
+    program_id: Pubkey,
+    expected_last_restart_slot: u64,
+) {
+    let instructions = vec![Instruction::new_with_bincode(
+        program_id,
+        &expected_last_restart_slot.to_le_bytes(),
+        vec![AccountMeta::new(last_restart_slot::id(), false)],
+    )];
+
+    let transaction = Transaction::new_signed_with_payer(
+        &instructions,
+        Some(&context.payer.pubkey()),
+        &[&context.payer],
+        context.last_blockhash,
+    );
+
+    context
+        .banks_client
+        .process_transaction(transaction)
+        .await
+        .unwrap();
+}
+
+#[tokio::test]
+async fn get_sysvar_last_restart_slot() {
+    let program_id = Pubkey::new_unique();
+    let program_test = ProgramTest::new(
+        "sysvar_last_restart_slot_process",
+        program_id,
+        processor!(sysvar_last_restart_slot_process_instruction),
+    );
+
+    let mut context = program_test.start_with_context().await;
+
+    check_with_program(&mut context, program_id, 0).await;
+    context.warp_to_slot(40).unwrap();
+    context.register_hard_fork(41 as Slot);
+    check_with_program(&mut context, program_id, 0).await;
+    context.warp_to_slot(42).unwrap();
+    check_with_program(&mut context, program_id, 41).await;
+    // check for value lower than previous hardfork
+    context.register_hard_fork(40 as Slot);
```

Comment:
Consider the following test case scenario:
```rust
    let mut context = program_test.start_with_context().await;

    check_with_program(&mut context, program_id, 0).await;
    context.warp_to_slot(39).unwrap();
    context.register_hard_fork(20 as Slot);
    context.warp_to_slot(40).unwrap();

    check_with_program(&mut context, program_id, 20).await;
```
- For the validator codepath, `bank_fork_utils::load_bank_forks()` ignores hard forks that are <= the snapshot slot (or 0 if we were loading from genesis).
- The `HardForks` struct itself does not impose such a restriction, and I believe 20 would be registered just fine. This would then lead to 20 being set in last_restart_slot variable.

I'm not sure this difference is a concrete problem, but at the very least, it could definitely be confusing. One option to enforce consistent behavior would be passing the bank (or at least its' slot) into `HardForks::register()`. Any thoughts on this @mvines ?

Reply:
`context.register_hard_fork(20 as Slot);` to me is nonsensical in this context. Actually panicking on an attempt to register a hard fork older than the working bank + 1 seems fine for example.

Reply:
Currently there is no panic when we attempt to register older hardfork. So we can always set an older hardfork.
Adding a panic seems to be out of context for this PR and may add some more issues, but happy to create a new issue for that.

https://github.com/solana-labs/solana/blob/47ff3cecc9cad527b208836f39a7321bf0511b5e/sdk/src/hard_forks.rs#LL18C8-L18C8

Reply:
thanks, i think it's fine to not add an explicit panic in that case